File: app/src/main/java/com/kabirnayeem99/paymentpaid/ui/activities/WorkDetailsActivity.java
package com.kabirnayeem99.paymentpaid.ui.activities;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.widget.DatePicker;
import android.widget.Toast;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;

import com.google.android.material.textfield.TextInputLayout;
import com.kabirnayeem99.paymentpaid.R;
import com.kabirnayeem99.paymentpaid.data.db.entities.Work;
import com.kabirnayeem99.paymentpaid.ui.WorkViewModel;
import com.kabirnayeem99.paymentpaid.utils.CustomUtils;

import static java.util.Objects.requireNonNull;

public class WorkDetailsActivity extends AppCompatActivity {

    public static final String EXTRA_ID = "com.kabirnayeem99.paymentpaid.ui.activities.EXTRA_ID";
    public static final String EXTRA_WORK_NAME = "com.kabirnayeem99.paymentpaid.ui.activities.EXTRA_WORK_NAME";
    public static final String EXTRA_PAYMENT = "com.kabirnayeem99.paymentpaid.ui.activities.EXTRA_PAYMENT";
    public static final String EXTRA_STUDENT_NAME = "com.kabirnayeem99.paymentpaid.ui.activities.EXTRA_STUDENT_NAME";
    public static final String EXTRA_DATE = "com.kabirnayeem99.paymentpaid.ui.activities.EXTRA_DATE";
    private static final String TAG = "WorkDetailsActivity";
    TextInputLayout tilWorkName;
    TextInputLayout tilPayment;
    TextInputLayout tilStudentName;
    DatePicker dpDate;
    Work work;
    String workName;
    String studentName;
    String date = String.format("%s-%s-%s", CustomUtils.getCurrentYear(),
            CustomUtils.padMonth(CustomUtils.getCurrentMonth() + 1),
            CustomUtils.padMonth(CustomUtils.getCurrentDay()));
    String paymentString;
    int month = CustomUtils.getCurrentMonth() + 1;
    int year = CustomUtils.getCurrentYear();
    Intent intent;
    int id = -1;
    WorkViewModel workViewModel;

    @RequiresApi(api = Build.VERSION_CODES.O)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_new_work);
        initViews();

        workViewModel = ViewModelProviders.of(WorkDetailsActivity.this).get(WorkViewModel.class);

        Log.d(TAG, "onCreate: default date" + date);
        assert dpDate != null;
        dpDate.setOnDateChangedListener((view, year, monthOfYear, dayOfMonth)
                -> {
            this.year = year;
            this.month = monthOfYear + 1;
            // month is returning a value less than the actual value, so magic number 1 is added
            date = String.format("%s-%s-%s", year, CustomUtils.padMonth(monthOfYear + 1),
                    CustomUtils.padMonth(dayOfMonth));
        });

        // manipulate views

        intent = getIntent();
        if (intent.hasExtra(EXTRA_ID)) {
            requireNonNull(tilWorkName.getEditText()).setText(intent.getStringExtra(EXTRA_WORK_NAME));
            tilWorkName.getEditText().setText(intent.getStringExtra(EXTRA_WORK_NAME));
            requireNonNull(tilPayment.getEditText()).setText(String.valueOf(intent.getIntExtra(EXTRA_PAYMENT, 0)));
            requireNonNull(tilStudentName.getEditText()).setText(intent.getStringExtra(EXTRA_STUDENT_NAME));
            String date = intent.getStringExtra(EXTRA_DATE);
        }
    }

    private void initViews() {
        tilWorkName = findViewById(R.id.til_work_name_add_new_work);
        tilPayment = findViewById(R.id.til_work_payment_add_new_work);
        tilStudentName = findViewById(R.id.til_student_name_add_new_work);
        dpDate = findViewById(R.id.dp_date_add_new_work);
    }


    @Override
    public void onBackPressed() {
        super.onBackPressed();
        saveToNoteDB();
    }

    private void saveToNoteDB() {
        // saves added note to the note database


        // requireNonNull ensures that the field is guaranteed be non-null.
        workName = requireNonNull(tilWorkName.getEditText()).getText().toString();
        studentName = requireNonNull(tilStudentName.getEditText()).getText().toString();
        paymentString = requireNonNull(tilPayment.getEditText()).getText().toString();


        if (!workName.trim().isEmpty() && !paymentString.trim().isEmpty() && !studentName.trim().isEmpty() && !date.trim().isEmpty()) {
            work = new Work(workName, date, month, year, Integer.parseInt(paymentString), studentName);
            id = intent.getIntExtra(EXTRA_ID, -1);
            if (id == -1) {
                Toast.makeText(this, "Work can't be update", Toast.LENGTH_SHORT).show();
                workViewModel.insert(work);
                Toast.makeText(WorkDetailsActivity.this, "Work is saved", Toast.LENGTH_SHORT).show();
            } else {
                work.setId(id);
                workViewModel.update(work);
            }

        } else {
            Toast.makeText(this, "Work was not saved", Toast.LENGTH_SHORT).show();
        }

    }
}

File: app/src/main/java/com/kabirnayeem99/paymentpaid/ui/activities/WorkDetailsActivity.kt
package com.kabirnayeem99.paymentpaid.ui.activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.google.firebase.auth.FirebaseAuth
import com.kabirnayeem99.paymentpaid.R
import com.kabirnayeem99.paymentpaid.data.db.entities.Work
import com.kabirnayeem99.paymentpaid.ui.*
import com.kabirnayeem99.paymentpaid.utils.CustomUtils
import kotlinx.android.synthetic.main.activity_add_new_work.*


/**
 * This Activity is for creating a new work or edit an existing work.
 * This [WorkDetailsActivity] extends [AppCompatActivity]
 */
class WorkDetailsActivity : AppCompatActivity() {
    private lateinit var firestoreViewModel: FirestoreViewModel
    private lateinit var work: Work

    private var toBeUpdatedWork: Work? = null
    val auth = FirebaseAuth.getInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_new_work)
        setUpUpdateWorkEditing()
        setUpViewModel()
    }


    /**
     * This method checks if any work has been passed down,
     * if not, it proceeds to fill the text field by calling
     * the [fillTheFieldIfExistingWork] function
     */
    private fun setUpUpdateWorkEditing() {
        toBeUpdatedWork = getWorkFromIntent()
        toBeUpdatedWork?.documentId?.let { fillTheFieldIfExistingWork(toBeUpdatedWork!!) }
    }

    /**
     * It gets the work that is passed down from the work list
     */
    private fun getWorkFromIntent(): Work? {
        val intent: Intent = intent

        val bundle = intent.extras
        return bundle?.getSerializable("work") as? Work

    }

    /**
     * This function fills the text fields if this work is a work
     * to be updated
     * @param workBeingProcessed which is a work object
     */
    private fun fillTheFieldIfExistingWork(workBeingProcessed: Work) {
        tilWorkName.editText?.setText(workBeingProcessed.name)
        tilStudentName.editText?.setText(workBeingProcessed.studentName)
        tilPayment.editText?.setText(workBeingProcessed.payment.toString())
        dpDate.updateDate(workBeingProcessed.year.toInt(), workBeingProcessed.month.toInt(), workBeingProcessed.day.toInt())
    }

    /**
     * overrides on back press function
     * with the back function, it checks if the work is a new work
     * or a work that already is exists
     * if new work it proceeds to insert it into the database
     * if existing work it proceeds to update the work in the database
     */
    override fun onBackPressed() {

        createOrUpdateWork()?.let { work ->
            if (toBeUpdatedWork?.documentId != null) {
                val success = firestoreViewModel.saveWork(work)
            } else {
                firestoreViewModel.saveWork(work)
            }
            Toast.makeText(this, "Your work was saved.", Toast.LENGTH_SHORT).show()
        }
        super.onBackPressed()
//        (this as Activity).overridePendingTransition(R.anim.slide_out_right, R.anim.fade_exit)
    }

    override fun overridePendingTransition(enterAnim: Int, exitAnim: Int) {
        super.overridePendingTransition(R.anim.slide_in_left, R.anim.slide_out_right)
    }

    /**
     * This methods sets up the view model for this activity
     */
    private fun setUpViewModel() {
        lateinit var logInRegisterViewModel: LogInRegisterViewModel
        val logInRegisterViewModelFactory = LogInRegisterViewModelProviderFactory(application)

        logInRegisterViewModel = ViewModelProvider(this,
                logInRegisterViewModelFactory).get(LogInRegisterViewModel::class.java)

        when (logInRegisterViewModel.getOfflineStatus()) {
            true -> {

                firestoreViewModel = ViewModelProviders.of(this).get(FirestoreViewModel::class.java)
            }
            false -> {
                firestoreViewModel = ViewModelProviders.of(this).get(FirestoreViewModel::class.java)
            }
        }

    }

    /**
     * It creates a new work or updates a work, based on the work type passed down
     * from the previous activity
     * @return Work? which can be null, if it fails to create work object
     */
    private fun createOrUpdateWork(): Work? {
        val workName: String = tilWorkName.editText?.text.toString()
        val studentName: String = tilStudentName.editText?.text.toString()
        val paymentAmount: String = tilPayment.editText?.text.toString()

        var month = CustomUtils.currentMonth
        var year = CustomUtils.currentYear
        var day = CustomUtils.currentDay

        dpDate?.let { dpDate ->
            month = dpDate.month + 1
            Log.d(TAG, "createOrUpdateWork: month $month")
            year = dpDate.year
            day = dpDate.dayOfMonth
        }

        /*
        returns null if any of the field is empty
        or returns a newly created work object
         */
        return if (workName.trim().isEmpty()
                || studentName.trim().isEmpty()
                || paymentAmount.trim().isEmpty()) {
            null
        } else {
            work = Work(documentId = auth.currentUser.uid + "3495", name = workName, day = day.toLong(), month = month.toLong(),
                    year = year.toLong(), payment = paymentAmount.toLong(), studentName = studentName.toString())



            toBeUpdatedWork?.let { workToBeUpdated ->
                work.documentId = workToBeUpdated.documentId
            }

            work
        }

    }

}


